Zookeeper certificate manager

stefan-chivu opened the pull request.

Commit:
Added method for updating gateway configuration in zookeeper loader and set watches for /certificates path in zookeeper

Commit:
Parametrized global cert path in GetZookeeperConfig

Commit:
Added ReconnectAll flag to TargetConnectionControl and changed GetZookeeperConfig() to refreshZookeeperConfig()

Commit:
Added reconnect logic when TLS certificates are added to refresh all existing connections with TLS

Commit:
Modified reconectAll logic

Commit:
changed statsd exporter metric logs to debug level

Commit:
Added dockerfile for statsd exporter example

Commit:
Created mechanism for setting watches for requests and fixed some minor issues with certificate auth

stefan-chivu committed:
Merge pull request #12 from stefan-chivu/zookeeper-request-monitoring

Created mechanism for setting watches for requests

petrutlucian94 commented on gateway/loaders/zookeeper/utils.go:
```diff
@@ -17,6 +22,55 @@ type ZookeeperClient struct {
 	conn *zk.Conn
 }
 
+func (z *ZookeeperTargetLoader) refreshZookeeperConfig(log zerolog.Logger, targetChan chan<- *connections.TargetConnectionControl, config *configuration.GatewayConfig) error {
```
nit: the method name is a bit confusing. The target config is also stored in zookeeper. How about "refreshGlobalConfig"?

petrutlucian94 commented on gateway/loaders/zookeeper/utils.go:
```diff
@@ -17,6 +22,55 @@ type ZookeeperClient struct {
+		controlMsg.ReconnectAll = true
+		targetChan <- controlMsg
+	} else {
+		config.Log.Info().Msg("No TLS Certificates found")
```
nit: let's log something like "No global config found", we may eventually use it for settings other than TLS certificates.

stefan-chivu replied:
the check is performed for the TLS config path, specifically /certificates/global

petrutlucian94 replied:
ohh, my bad, I think I misread that

petrutlucian94 commented on gateway/loaders/zookeeper/zookeeper.go:
```diff
@@ -163,24 +176,20 @@ func (z *ZookeeperTargetLoader) WatchConfiguration(targetChan chan<- *connection
-		} else {
-			z.config.Log.Info().Msgf("[ZK] Successfuly set data watch for: %s", watchPath)
+	// TODO: Fix reconnect issue for target connections
+	// if err := z.refreshZookeeperConfig(z.config.Log, targetChan, z.config); err != nil {
```
leftover?

petrutlucian94 reviewed (commented):
just a few minor nits. apart from that, we'll need to fix the target reconnect workflow.

petrutlucian94 commented on gateway/connections/zookeeper.go:
```diff
@@ -187,6 +187,64 @@ func (c *ZookeeperConnectionManager) handleTargetControlMsg(msg *TargetConnectio
+		for _, conn := range c.connections {
+			// c.config.Log.Info().Msgf("Reconnecting conn: ", conn)
+			if err := conn.reconnect(c.connLimit); err != nil {
+				c.config.Log.Error().Msgf("Recconect failed for connection: ", err)
```
nit: s/Recconect/Reconnect. also, "Reconnect failed" is enough, "Reconnect failed for connection" sounds a bit awkward.

Commit:
Refactored refreshZookeeperConfig to refreshGlobalConfig and other minor changes

petrutlucian94 committed:
Avoid deadlock when reconnecting

The gnmi cache gets reset when removing or recreating connections.
Any cached notification will be pushed to the exporters.

The issue is that ZookeeperConnectionManager.Forwardable loops
over the connections, tries to acquire the connection lock and
enters a deadlock.

A simple workaround would be to just invoke the disconnect/reconnect
calls asynchronously. Worth mentioning that "connect" call is already
async.

At the same time, we can restore the old reconnect implementation,
which is simply closing the underlying connection, which in turn will
trigger a reconnect attempt.

stefan-chivu committed:
Merge pull request #13 from petrutlucian94/zookeeper-certificates

Avoid deadlock when reconnecting

petrutlucian94 commented:
I think this is ready to go now.

Commit:
Added basic request update logic for connections

Commit:
Removed unused parameter from reconnect()

stefan-chivu committed:
Merge pull request #14 from stefan-chivu/zookeeper-request-monitoring

Request update logic for connections
